frontend/pages/hosts/details/cards/Software/SelfService/SelfService.tsx

```diff
@@ -235,6 +313,96 @@ const SoftwareSelfService = ({
     refetchForPendingInstallsOrUninstalls();
   }, [refetchForPendingInstallsOrUninstalls]);
 
+  const isMountedRef = useRef(false);
+  useEffect(() => {
+    isMountedRef.current = true;
+    return () => {
+      isMountedRef.current = false;
+    };
+  }, []);
+
+  const onClickInstallAction = useCallback(
+    async (softwareId: number) => {
+      try {
+        await deviceApi.installSelfServiceSoftware(deviceToken, softwareId);
+        if (isMountedRef.current) {
+          onInstallOrUninstall();
+        }
+      } catch (error) {
+        // We only show toast message if API returns an error
+        renderFlash("error", "Couldn't install. Please try again.");
+      }
+    },
+    [deviceToken, onInstallOrUninstall, renderFlash]
+  );
+
+  const onClickUpdateAction = useCallback(
+    async (id: number) => {
+      try {
+        await deviceApi.installSelfServiceSoftware(deviceToken, id);
+        onInstallOrUninstall();
+      } catch (error) {
+        // Only show toast message if API returns an error
+        renderFlash("error", "Couldn't update software. Please try again.");
+      }
+    },
+    [deviceToken, onInstallOrUninstall, renderFlash]
+  );
+
+  const onClickUpdateAll = useCallback(async () => {
```

Comment:
just checking there is a loading state for when button to update all the software is clicked correct? I don't see code in here to set a loading state but I imagine that's handled at the software card level?

Reply:
Yeah, it updates the returned API statuses to `pending_install` which updates `ui_status` which controls the card states